src/joystick/hidapi/SDL_hidapi_sinput.c

```diff
@@ -0,0 +1,798 @@
+}
+
+static bool HIDAPI_DriverSInput_IsSupportedDevice(SDL_HIDAPI_Device *device, const char *name, SDL_GamepadType type, Uint16 vendor_id, Uint16 product_id, Uint16 version, int interface_number, int interface_class, int interface_subclass, int interface_protocol)
+{
+    return SDL_IsJoystickSInputController(vendor_id, product_id);
+}
+
+static bool HIDAPI_DriverSInput_InitDevice(SDL_HIDAPI_Device *device)
+{
+#if defined(DEBUG_SINPUT_INIT)
+    SDL_Log("SInput device Init");
+#endif
+
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)SDL_calloc(1, sizeof(*ctx));
+    if (!ctx) {
+        return false;
+    }
+
+    ctx->device = device;
+    device->context = ctx;
+
+    if (!RetrieveSDLFeatures(device)) {
+        return false;
+    }
+    
+    return HIDAPI_JoystickConnected(device, NULL);
+}
+
+static int HIDAPI_DriverSInput_GetDevicePlayerIndex(SDL_HIDAPI_Device *device, SDL_JoystickID instance_id)
+{
+    return -1;
+}
+
+static void HIDAPI_DriverSInput_SetDevicePlayerIndex(SDL_HIDAPI_Device *device, SDL_JoystickID instance_id, int player_index)
+{
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)device->context;
+
+    if (ctx->player_leds_supported) {
+        player_index = SDL_clamp(player_index + 1, 0, 255);
+        Uint8 player_num = (Uint8)player_index;
+
+        ctx->player_idx = player_num;
+
+        // Set player number, finalizing the setup
+        Uint8 playerLedCommand[SINPUT_DEVICE_REPORT_COMMAND_SIZE] = { SINPUT_DEVICE_REPORT_ID_OUTPUT_CMDDAT, SINPUT_DEVICE_COMMAND_PLAYERLED, ctx->player_idx };
+        int playerNumBytesWritten = SDL_hid_write(device->dev, playerLedCommand, SINPUT_DEVICE_REPORT_COMMAND_SIZE);
+
+        if (playerNumBytesWritten < 0) {
+            SDL_SetError("SInput device player led command could not write");
+        }
+    }
+}
+
+#ifndef DEG2RAD
+#define DEG2RAD(x) ((float)(x) * (float)(SDL_PI_F / 180.f))
+#endif
+
+
+static bool HIDAPI_DriverSInput_OpenJoystick(SDL_HIDAPI_Device *device, SDL_Joystick *joystick)
+{
+#if defined(DEBUG_SINPUT_INIT)
+    SDL_Log("SInput device Open");
+#endif
+
+    SDL_DriverSInput_Context *ctx = (SDL_DriverSInput_Context *)device->context;
+
+    SDL_AssertJoysticksLocked();
+
+    ctx->joystick = joystick;
+
+    joystick->nbuttons = ctx->buttons_count;
+
+    SDL_zeroa(ctx->last_state);
+
+    int axes = 0;
+    if (ctx->left_analog_stick_supported) {
+        axes += 2;
+    }
+
+    if (ctx->right_analog_stick_supported) {
+        axes += 2;
+    }
+
+    if (ctx->left_analog_trigger_supported || ctx->left_digital_trigger_supported) {
+        ++axes;
+    }
+
+    if (ctx->right_analog_trigger_supported || ctx->right_digital_trigger_supported) {
+        ++axes;
+    }
+
+    joystick->naxes = axes;
+
+    if (ctx->accelerometer_supported) {
+        SDL_PrivateJoystickAddSensor(joystick, SDL_SENSOR_ACCEL, (float)SINPUT_DEVICE_POLLING_RATE_MS);
```

Comment:
You might want to add the polling rate to the info returned by the device. Different devices may have different polling rates.